File: LeetCode/514. Freedom Trail/514.py
class Solution(object):
    def findRotateSteps(self, ring, key):
        """
        :type ring: str
        :type key: str
        :rtype: int
        """
        def find_positions(ring, character):
            positions = []
            for i, char in enumerate(ring):
                if char == character:
                    positions.append(i)
            return positions
        
        def min_steps_to_position(ring, current_pos, target_pos):
            clockwise_steps = (target_pos - current_pos) % len(ring)
            anticlockwise_steps = (current_pos - target_pos) % len(ring)
            return min(clockwise_steps, anticlockwise_steps)
        
        def dp(ring_index, key_index, memo):
            if key_index == len(key):
                return 0
            
            if (ring_index, key_index) in memo:
                return memo[(ring_index, key_index)]
            
            min_steps = float('inf')
            for pos in find_positions(ring, key[key_index]):
                steps_to_position = min_steps_to_position(ring, ring_index, pos)
                steps_to_next_char = dp(pos, key_index + 1, memo)
                total_steps = steps_to_position + steps_to_next_char
                min_steps = min(min_steps, total_steps)
                
            memo[(ring_index, key_index)] = min_steps + 1  # Ekstra bir adım merkez düğmesine basmak için
            return min_steps + 1  # Bir sonraki karakteri işaretlemeden önce merkez düğmesine basılması gerektiği için +1
        
        return dp(0, 0, {})


#2
class Solution(object):
    def findRotateSteps(self, ring, key):
        """
        :type ring: str
        :type key: str
        :rtype: int
        """
        n, m = len(ring), len(key)
        dp = [[float('inf')] * n for _ in range(m)]
        pos = [[] for _ in range(26)]
        for i in range(n):
            pos[ord(ring[i]) - ord('a')].append(i)
        for i in pos[ord(key[0]) - ord('a')]:
            dp[0][i] = min(i, n - i) + 1
        for i in range(1, m):
            for j in pos[ord(key[i]) - ord('a')]:
                for k in pos[ord(key[i - 1]) - ord('a')]:
                    diff = abs(j - k)
                    dp[i][j] = min(dp[i][j], dp[i - 1][k] + min(diff, n - diff) + 1)
        return min(dp[m - 1])

File: LeetCode/514. Freedom Trail/514.java
import java.util.HashMap;
import java.util.Map;

class Solution {
    public int findRotateSteps(String ring, String key) {
        return dp(ring, key, 0, 0, new HashMap<>());
    }
    
    private int dp(String ring, String key, int ringIndex, int keyIndex, Map<String, Integer> memo) {
        if (keyIndex == key.length()) {
            return 0;
        }
        
        String memoKey = ringIndex + "_" + keyIndex;
        if (memo.containsKey(memoKey)) {
            return memo.get(memoKey);
        }
        
        char targetChar = key.charAt(keyIndex);
        int minSteps = Integer.MAX_VALUE;
        for (int pos : findPositions(ring, targetChar)) {
            int stepsToPosition = minStepsToPosition(ring, ringIndex, pos);
            int stepsToNextChar = dp(ring, key, pos, keyIndex + 1, memo);
            int totalSteps = stepsToPosition + stepsToNextChar;
            minSteps = Math.min(minSteps, totalSteps);
        }
        
        memo.put(memoKey, minSteps + 1); // Merkez düğmesine basmak için ekstra bir adım
        return minSteps + 1; // Bir sonraki karakteri işaretlemeden önce merkez düğmesine basılması gerektiği için +1
    }
    
    private int minStepsToPosition(String ring, int currentPos, int targetPos) {
        int clockwiseSteps = Math.abs(targetPos - currentPos);
        int anticlockwiseSteps = ring.length() - clockwiseSteps;
        return Math.min(clockwiseSteps, anticlockwiseSteps);
    }
    
    private int[] findPositions(String ring, char character) {
        int count = 0;
        for (char c : ring.toCharArray()) {
            if (c == character) {
                count++;
            }
        }
        int[] positions = new int[count];
        int index = 0;
        for (int i = 0; i < ring.length(); i++) {
            if (ring.charAt(i) == character) {
                positions[index++] = i;
            }
        }
        return positions;
    }
}

//2.cozum
class Solution {
    public int findRotateSteps(String ring, String key) {
          char[] r=ring.toCharArray();
        List<Integer>[] p=new List[26];
        for(int i=0;i<r.length;i++) {
            int c=r[i]-'a';
            List<Integer> l=p[c];
            if(l==null) p[c]=l=new ArrayList<>();
            l.add(i);
        }
        return helper(0,0,p,key.toCharArray(),ring,new int[key.length()][r.length]);
    }
    int helper(int in, int pos, List<Integer>[] p, char[] k, String r, int[][] memo) {
        if(in==k.length) return 0;
        if(memo[in][pos]>0) return memo[in][pos]-1;
        int min=Integer.MAX_VALUE;
        for(int i: p[k[in]-'a']) {
            int m;
            if(i>=pos) m=Math.min(i-pos,pos+r.length()-i);
            else m=Math.min(pos-i,i+r.length()-pos);
            min=Math.min(min,m+helper(in+1,i,p,k,r,memo));
        }
        return (memo[in][pos]=min+2)-1;
    }
}